### PR TITLE
Fix race condition between validating a candidate hash, and accepting a candidate block

### LINKED ISSUE #263: Race condition between validating a candidate hash and accepting a candidate message
As observed on the testnet, often times the candidate message will arrive to the candidate broker before the block hash is validated by the selector. This leads to unnecessary delays in reaching consensus.

### LINKED ISSUE #264: Candidate broker `validHashes` map should be thread-safe
Currently, the `validHashes` map is not protected for concurrent use, though it is used across multiple goroutines, as hashes are validated through a callback on the event bus,

### LINKED ISSUE #267: Mempool should be cleared both on the accepted block update, and the intermediate block update
On the testnet, some block generators were observed attempting to include the same transactions twice. This happened whenever the block generator went into sleeper mode and had to catch up a few blocks. Since the mempool is only cleared on the intermediate block, these transactions which should have been cleaned up, remain in their mempool, and inevitably find their way into a block.

### PR DESCRIPTION
**Pull request by julesdesmit**
Fixes #263, fixes #264, fixes #267

### CODE REVIEW

File: pkg/util/nativeutils/republisher/republisher.go [hunk 1]
```diff
@@ -5,9 +5,10 @@ import (
 
 	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/topics"
 	"github.com/dusk-network/dusk-blockchain/pkg/util/nativeutils/eventbus"
+	"github.com/dusk-network/dusk-blockchain/pkg/util/nativeutils/prerror"
 )
 
-type Validator func(bytes.Buffer) error
+type Validator func(bytes.Buffer) *prerror.PrError
```
Comment (autholykos) [inline pkg/util/nativeutils/republisher/republisher.go]: I somewhat dislike setting priorities during `error` creation. I get the impression that it is a practice that pollutes the interface by forcing implementers to handle information they should have no say about. Whether an error is high priority or not, is none of the Validator business, and it should be a responsibility left to the components that handle the validation errors. I would suggest to include some common validation errors (something like "DuplicatePayloadError", "EncodingError" and "InvalidError") and type-check/handle them at the caller level

File: pkg/core/candidate/broker.go [hunk 2]
```diff
@@ -27,13 +26,14 @@ type Broker struct {
 	publisher   eventbus.Publisher
 	republisher *republisher.Republisher
 	*store
-	// List of block hashes for which a valid Score message was seen.
 	lock        sync.RWMutex
+	queue       map[string]*Candidate
```
[the remark this answers is not in the document]
Comment (julesdesmit) [reply pkg/core/candidate/broker.go]: Good point, will adjust it.

File: pkg/util/nativeutils/republisher/republisher.go (hunk 1 above)
Comment (julesdesmit) [reply pkg/util/nativeutils/republisher/republisher.go]: I see what you mean. I'll adjust it.

File: pkg/util/nativeutils/republisher/republisher.go [hunk 3]
```diff
@@ -52,15 +60,9 @@ func (r *Republisher) Activate() uint32 {
 func (r *Republisher) Republish(b bytes.Buffer) error {
 	for _, v := range r.validators {
 		err := v(b)
-		if err != nil && err.Priority == prerror.High {
+		if err != nil && (err == InvalidError || err == EncodingError) {
```
Comment (autholykos) [inline pkg/util/nativeutils/republisher/republisher.go]: nitpicking, but I'd rather use a `switch` with a `default` rather than an `if`

**Pull request merged**